fish-rust/src/env_dispatch.rs

```diff
@@ -0,0 +1,809 @@
+}
+
+fn handle_locale_change(vars: &dyn Environment) {
+    init_locale(vars);
+    // We need to re-guess emoji width because the locale might have changed to a multibyte one.
+    guess_emoji_width(vars);
+}
+
+fn handle_curses_change(vars: &dyn Environment) {
+    guess_emoji_width(vars);
+    init_curses(vars);
+}
+
+fn handle_fish_use_posix_spawn_change(vars: &dyn Environment) {
+    // Note that if the variable is missing or empty we default to true (if allowed).
+    if !allow_use_posix_spawn() {
+        USE_POSIX_SPAWN.store(false, Ordering::Relaxed);
+    } else if let Some(var) = vars.get(L!("fish_use_posix_spawn")) {
+        let use_posix_spawn =
+            var.is_empty() || crate::wcstringutil::bool_from_string(&var.as_string());
+        USE_POSIX_SPAWN.store(use_posix_spawn, Ordering::Relaxed);
+    } else {
+        USE_POSIX_SPAWN.store(true, Ordering::Relaxed);
+    }
+}
+
+/// Allow the user to override the limits on how much data the `read` command will process. This is
+/// primarily intended for testing, but could also be used directly by users in special situations.
+fn handle_read_limit_change(vars: &dyn Environment) {
+    let read_byte_limit = vars
+        .get_unless_empty(L!("fish_read_limit"))
+        .map(|v| v.as_string())
+        .and_then(|v| match v.to_string().parse() {
+            Ok(v) => Some(v),
+            Err(_) => {
+                // XXX: In the C++ code, this warning wasn't behind an "is_interactive_session()"
+                // check.
+                if is_interactive_session() {
+                    FLOGF!(warning, "Ignoring invalid $fish_read_limit");
+                }
+                None
+            }
+        })
+        .unwrap_or(DEFAULT_READ_BYTE_LIMIT);
+    READ_BYTE_LIMIT.store(read_byte_limit, Ordering::Relaxed);
+}
+
+fn handle_fish_trace(vars: &dyn Environment) {
+    let enabled = vars.get_unless_empty(L!("fish_trace")).is_some();
+    crate::trace::trace_set_enabled(enabled);
+}
+
+pub fn env_dispatch_init(vars: &EnvStack) {
+    use once_cell::sync::Lazy;
+
+    run_inits(vars);
+    // env_dispatch_var_change() purposely supresses change notifications until the dispatch table
+    // was initialized elsewhere (either explicitly as below or via deref of VAR_DISPATCH_TABLE).
+    Lazy::force(&VAR_DISPATCH_TABLE);
+}
+
+pub fn env_dispatch_init_ffi() {
+    let vars = EnvStack::principal();
+    env_dispatch_init(vars);
+}
+
+/// Runs the subset of dispatch functions that need to be called at startup.
+fn run_inits(vars: &dyn Environment) {
+    init_locale(vars);
+    init_curses(vars);
+    guess_emoji_width(vars);
+    update_wait_on_escape_ms(vars);
+    handle_read_limit_change(vars);
+    handle_fish_use_posix_spawn_change(vars);
+    handle_fish_trace(vars);
+}
+
+/// Updates our idea of whether we support term256 and term24bit (see issue #10222).
+fn update_fish_color_support(vars: &dyn Environment) {
+    let max_colors = curses::term().get(curses::MAX_COLORS);
```

Comment:
The previous code guarded this with `cur_term != nullptr`; it looks like this will crash if cur_term is not set?

Reply:
Yup, that was lost in the rewrite. Thanks for catching. Changing `curses::term()` to return an `Option` is certainly going to be the safest approach.

Reply:
I have been porting this check as `int has_cur_term() { return cur_term != NULL; }`

Reply:
@krobelus We'll want to unify the "uses curses" code between our forks. I have a `curses::is_initialized()` function that does that check but with the most recent commits that change `curses::term()` to return an `Option<&'static Term>` we wouldn't actually use it anywhere. All curses functionality should be restricted to `curses.rs` and all functionality requiring a valid `cur_term` ptr should be restricted to methods on the `Term` type which is only available if `cur_term != nullptr`.